### PR TITLE
Support converting Iceberg for CONVERT TO DELTA command in Delta Lake

### PR DESCRIPTION
**Pull request by jackierwzhang**
## Description
Adding support for **in-place** converting Iceberg to Delta using the CONVERT TO DELTA command in Apache Spark. Specifically, this PR supports converting a Parquet-based Iceberg table inside a path/directory to Delta Lake format.

Here's an example flow:
1. Given a Spark environment
2. Follow the Iceberg setup [here](https://iceberg.apache.org/docs/latest/spark-configuration/#catalogs). **Please use the hadoop directory based catalog** so we could find Iceberg in a path.
3. Suppose now we have an iceberg table sitting inside `s3://bucket/catalog/db/table`
4. Run the following command
```
CONVERT TO DELTA iceberg.`s3://bucket/catalog/db/table`
```
5. Now you have a Delta table at the same location!
6. To bring this Delta table into any Spark catalog, simply run `CREATE TABLE delta_table USING delta LOCATION 's3://bucket/catalog/db/table'`

See more detail in this ticket: https://github.com/delta-io/delta/issues/1462.

## How was this patch tested?
New unit tests.

We have tested Iceberg version from 0.13.1 to 1.0.0.

## Does this PR introduce _any_ user-facing changes?
It introduces a `iceberg-delta-compat` module that contains all the Iceberg + Spark dependencies, please include this module during Spark startup so CONVERT TO DELTA command could work.

### CODE REVIEW

**Comment by scottsand-db** [comment]
@jackierwzhang 

> Does this PR introduce any user-facing changes?

This section should also discuss the new maven artifact your PR is proposing, and how to include it in a spark session along with delta lake.

Also - have you tested this on a real cluster (e.g. EMR)? Your `build.sbt` project definition is a little unlike the others. e.g. yours includes `.cross(CrossVersion.binary)` yet `commonSettings` already includes `crossScalaVersions := all_scala_versions,`

**Comment by jackierwzhang** [comment]
> Also - have you tested this on a real cluster (e.g. EMR)? 

Yes, I have tested this on a real cluster.

> Your `build.sbt` project definition is a little unlike the others. e.g. yours includes `.cross(CrossVersion.binary)` yet `commonSettings` already includes `crossScalaVersions := all_scala_versions,`, so I'd want to see this tested.

Yes, so basically this will allow the scala build to cross compile different iceberg versions against `all_scala_versions` as we defined.

> we should add an integration test for this. This PR is already large enough so I think we can add this in a future PR.

Sure but imo the unit test is already a complete Spark environment with a standard Iceberg setup.

File: build.sbt [hunk 1]
```diff
@@ -194,6 +194,21 @@ lazy val storageS3DynamoDB = (project in file("storage-s3-dynamodb"))
     )
   )
 
+lazy val deltaIcebergCompat = (project in file("delta-iceberg-compat"))
+  .dependsOn(core % "compile->compile;test->test;provided->provided")
+  .settings (
+    name := "delta-iceberg-compat",
```
Comment (tdas) [inline build.sbt:200]: @zsxwing do you think this is a good name? 
compat or compatibility? 

and should we have "spark" in the name?

**Comment by scottsand-db** [comment]
> > we should add an integration test for this. This PR is already large enough so I think we can add this in a future PR.

> Sure but imo the unit test is already a complete Spark environment with a standard Iceberg setup.

Yes but this doesn't test the packaged JAR. We've caught many bugs (e.g. shading, dependencies, scala version issues, etc.) before with our integration tests.

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaErrors.scala [hunk 2]
```diff
@@ -2452,6 +2452,11 @@ trait DeltaErrorsBase
         unsupportedOptions.mkString(","))
     )
   }
+
+  def partitionSchemaInIcebergTables: Throwable = {
+    new DeltaIllegalArgumentException(
+      errorClass = "DELTA_PARTITION_SCHEMA_IN_ICEBERG_TABLES", messageParameters = null)
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaErrors.scala:2458]: nit: There is already a default arg for `messageParameters`

File: core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala [hunk 3]
```diff
@@ -280,19 +283,22 @@ abstract class ConvertToDeltaCommandBase(
...
-          new ParquetTable(spark, qualifiedDir, target.catalogTable, partitionSchema)
-        case checkProvider =>
-          throw DeltaErrors.convertNonParquetTablesException(tableIdentifier, checkProvider)
+        case checkProvider
+          if target.catalogTable.exists(ConvertToDeltaCommand.isHiveStyleParquetTable) ||
+            checkProvider.equalsIgnoreCase("parquet") =>
+          ConvertToDeltaCommand.getParquetTable(
+            spark, target.targetDir, target.catalogTable, partitionSchema)
+        case checkProvider
+          if icebergEnabled && checkProvider.equalsIgnoreCase("iceberg") =>
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:295]: nit: `icebergEnabled && checkProvider.equalsIgnoreCase("iceberg")` - utility method `isIcebergTabled`. Can be used in the `isSupportedProvider` as well.

File: core/src/main/resources/error/delta-error-classes.json [hunk 4]
```diff
@@ -1202,6 +1202,12 @@
     ],
     "sqlState" : "42000"
   },
+  "DELTA_PARTITION_SCHEMA_IN_ICEBERG_TABLES" : {
+    "message" : [
+      "Partition schema cannot be specified when converting Iceberg tables"
```
Comment (vkorukanti) [inline core/src/main/resources/error/delta-error-classes.json:1207]: Is the partition schema automatically derived? If yes, can you add the details here?

File: core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala [hunk 5]
```diff
@@ -841,4 +847,43 @@ object ConvertToDeltaCommand {
...
+
+  def getIcebergTable(
+      spark: SparkSession,
+      targetDir: String,
+      sparkTable: Option[Table],
+      tableSchema: Option[StructType]): ConvertTargetTable = {
+    try {
+      val clazz = Utils.classForName("org.apache.spark.sql.delta.IcebergTable")
+      if (sparkTable.isDefined) {
+        val constFromTable = clazz.getConstructor(
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:868]: nit: Catch the `method not found exception` and throw a proper error message if the required jars need to be present? This doesn't need to be done here, may be in a follow up PR.

**Review by vkorukanti — COMMENTED** [review]
+1 on adding an integration test.

Also docs on what jars needed to be copied (end-2-end steps)

Otherwise LGTM.

**Comment by vkorukanti** [comment]
Are there any limitations on what types/features in Iceberg supported and not supported?

File: core/src/main/resources/error/delta-error-classes.json (hunk 4 above)
Comment (jackierwzhang) [reply core/src/main/resources/error/delta-error-classes.json:1207]: yes, will do.

**Comment by jackierwzhang** [comment]
> Are there any limitations on what types/features in Iceberg supported and not supported?

There are a few major things we don't yet support:

1. Converting hive based iceberg table
2. Converting non-parquet iceberg table
3. Converting special partition transformation rules ([ref](https://github.com/delta-io/delta/pull/1463/files#diff-a5bb63f4073eeff36c84d70fc4e0b0b9191eaeed2b6885293da94a4cc5fa8d81R154))
4. Converting iceberg with custom name mapping ([ref](https://github.com/delta-io/delta/pull/1463/files#diff-ffa943e0f0f3dca25f431b29932a0ed62ca9a4d80b929e4a0a5e8b94cf6fc71dR132))
5. Converting iceberg table with case-sensitive column names ([ref](https://github.com/delta-io/delta/pull/1463/files#diff-ffa943e0f0f3dca25f431b29932a0ed62ca9a4d80b929e4a0a5e8b94cf6fc71dR144))

File: core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala (hunk 5 above)
Comment (jackierwzhang) [reply core/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala:868]: can do in a follow up.

**Review by vkorukanti — APPROVED** [review]
lgtm

File: core/src/main/resources/error/delta-error-classes.json [hunk 6]
```diff
@@ -946,6 +946,12 @@
     ],
     "sqlState" : "42000"
   },
+  "DELTA_MISSING_ICEBERG_CLASS" : {
+    "message" : [
+      "Iceberg class path is not found. Please ensure Delta Iceberg support is installed.",
```
Comment (tdas) [inline core/src/main/resources/error/delta-error-classes.json:951]:
```suggestion
      "Iceberg class was not found. Please ensure Delta Iceberg support is installed.",
```

**Pull request closed without merging**